crates/sui-core/src/authority_aggregator.rs

```diff
@@ -349,6 +353,31 @@ impl ProcessTransactionState {
         }
         false
     }
+
+    pub fn check_if_error_indicates_tx_finalized_with_different_user_sig(
+        &self,
+        quorum_threshold: StakeUnit,
+    ) -> bool {
+        // In some edge cases, the client may sent the same transaction multiple times but with different user signatures.
+        // When this happens, the "minority" tx will fail in safe_client because the certificate verification would fail
+        // and return Sui::FailedToVerifyTxCertWithExecutedEffects.
+        // Here, we check if there are 2f+1 validators return this error. If so, the transaction is already finalized
+        // with a different set of user signatures. It's not trivial to return the results of that successful transaction
+        // because we don't want fullnode to store the transaction with non-canonical user signatures. Given that this is
+        // very rare, we simply return an error here.
+        let invalid_sig_stake: StakeUnit = self
```

Comment:
no preferences from me, I can change it to that if you feel strongly

Reply:
No preference from me as well, will leave it up to you. 

I think I would only go with my suggestion if `FailedToVerifyTxCertWithExecutedEffects` is indeed a nonretryable error. If it is, then I was thinking you could calculate `invalid_sig_stake` on the fly here https://github.com/MystenLabs/sui/blob/59b4621428970f9a4a6d496ad31ce2ce3f73ec97/crates/sui-core/src/authority_aggregator.rs#L1198-L1262 Then `state.retryable` would be correctly set and you could just check `if !(state.invalid_sig_stake >= quorum_threshold)` before the panic. Which matches how the other special case errors are handled. Otherwise what you have makes sense to me